File: frontend/app/lib/sources.ts
/**
 * Browser-side sources client (T7, SRC-11).
 *
 * Thin helpers the sources screen uses to talk to FastAPI *through the
 * same-origin Next.js proxy* (`/api/...`, ADR-017) — never cross-origin. The
 * HttpOnly session cookie rides along automatically (`credentials:
 * "same-origin"`), so this code never reads or holds the session token. The
 * only token JS handles is the CSRF token (from `/api/auth/me`), echoed in
 * `X-CSRF-Token` on the state-changing upload (AD-007), mirroring `auth.ts`.
 *
 * FastAPI remains authoritative for auth, ownership, and validation; these
 * helpers just carry inputs in and surface success/error out.
 */

/** Secret-free source summary as returned by `/api/sources*`. */
export type SourceSummary = {
  id: string;
  title: string;
  filename: string;
  byte_size: number;
  content_type: string;
  status: string;
  created_at: string;
};

/** One progress-log entry, as returned in an `IngestionSummary`. */
export type IngestionEventView = {
  type: string;
  message: string | null;
  created_at: string;
};

/**
 * Secret-free ingestion job view as returned by `/api/sources/{id}/ingestion`.
 * Mirrors the backend `IngestionSummary`: job lifecycle state only, never the
 * source's `object_key`/`checksum`.
 */
export type IngestionSummary = {
  id: string;
  status: string;
  attempts: number;
  error: string | null;
  created_at: string;
  updated_at: string;
  events: IngestionEventView[];
};

/**
 * One node in a source's parsed section tree (CORP-11), nested per the TOC
 * hierarchy. Mirrors the backend `StructureSectionView`.
 */
export type StructureSection = {
  title: string;
  depth: number;
  section_path: string[];
  anchor: string;
  children: StructureSection[];
};

/**
 * A source's parsed book structure as returned by
 * `/api/sources/{id}/structure`. Mirrors the backend `BookStructureView`:
 * `title`/`language` are null and `authors` empty when the OPF omitted them.
 */
export type SourceStructure = {
  title: string | null;
  authors: string[];
  language: string | null;
  sections: StructureSection[];
};

/** List the caller's sources (newest-first), or `[]` when they have none. */
export async function listSources(
  fetchImpl: typeof fetch = fetch,
): Promise<SourceSummary[]> {
  const res = await fetchImpl("/api/sources", {
    method: "GET",
    credentials: "same-origin",
  });
  if (!res.ok) {
    throw await toSourceError(res, "Could not load your sources.");
  }
  return (await res.json()) as SourceSummary[];
}

/** Fetch one source the caller owns. */
export async function getSource(
  id: string,
  fetchImpl: typeof fetch = fetch,
): Promise<SourceSummary> {
  const res = await fetchImpl(`/api/sources/${id}`, {
    method: "GET",
    credentials: "same-origin",
  });
  if (!res.ok) {
    throw await toSourceError(res, "Could not load that source.");
  }
  return (await res.json()) as SourceSummary;
}

/**
 * Upload an EPUB with a title. This is a state-changing request, so it carries
 * the session-bound CSRF token in `X-CSRF-Token` (AD-007); the caller passes it
 * in (read from `/api/auth/me`), same as `logout(csrfToken)`. The body is real
 * multipart `FormData` (file + title) — we deliberately do NOT set a
 * `Content-Type`, letting the browser add the multipart boundary itself.
 */
export async function uploadSource(
  file: File,
  title: string,
  csrfToken: string,
  fetchImpl: typeof fetch = fetch,
): Promise<SourceSummary> {
  const body = new FormData();
  body.set("file", file);
  body.set("title", title);
  const res = await fetchImpl("/api/sources", {
    method: "POST",
    credentials: "same-origin",
    headers: { "X-CSRF-Token": csrfToken },
    body,
  });
  if (!res.ok) {
    throw await toSourceError(res, "Upload failed.");
  }
  return (await res.json()) as SourceSummary;
}

/**
 * Start ingestion for an uploaded source. This is a state-changing request, so
 * it carries the session-bound CSRF token in `X-CSRF-Token` (AD-007), same as
 * `uploadSource`; the caller passes it in (read from `/api/auth/me`). The
 * HttpOnly session cookie rides along automatically (`credentials:
 * "same-origin"`). On a non-OK response (e.g. 409 "already in progress", 502),
 * the backend `detail` is surfaced via `toSourceError`.
 */
export async function startIngestion(
  sourceId: string,
  csrfToken: string,
  fetchImpl: typeof fetch = fetch,
): Promise<IngestionSummary> {
  const res = await fetchImpl(`/api/sources/${sourceId}/ingestion`, {
    method: "POST",
    credentials: "same-origin",
    headers: { "X-CSRF-Token": csrfToken },
  });
  if (!res.ok) {
    throw await toSourceError(res, "Could not start ingestion.");
  }
  return (await res.json()) as IngestionSummary;
}

/**
 * Fetch a source's parsed book structure (metadata + nested section tree)
 * through the same-origin proxy. Read-only GET, so no CSRF token; the HttpOnly
 * session cookie rides along automatically. Non-OK responses (e.g. 404 when the
 * source has no corpus yet) surface the backend `detail` via `toSourceError`,
 * mirroring `listSources`.
 */
export async function fetchSourceStructure(
  id: string,
  fetchImpl: typeof fetch = fetch,
): Promise<SourceStructure> {
  const res = await fetchImpl(`/api/sources/${id}/structure`, {
    method: "GET",
    credentials: "same-origin",
  });
  if (!res.ok) {
    throw await toSourceError(res, "Could not load the book structure.");
  }
  return (await res.json()) as SourceStructure;
}

/** Build an Error from a non-OK response, preferring the backend's detail. */
async function toSourceError(res: Response, fallback: string): Promise<Error> {
  try {
    const body = (await res.json()) as { detail?: string };
    return new Error(body.detail ?? fallback);
  } catch {
    return new Error(fallback);
  }
}

File: backend/app/infrastructure/web/sources.py
"""Sources router — upload, list, and read owned EPUB sources (Cycle 2).

Thin FastAPI adapter over the framework-free source services (assembled in
``dependencies``). Handlers delegate to a use-case service and return a
secret-free ``SourceSummary``; application errors raised by the services are
translated to HTTP status codes by the global handlers in ``error_handlers``.

Contract (also consumed by the Next.js proxy in Phase 4):
- ``POST /api/sources``        → 201, multipart ``file`` + ``title``; auth + CSRF.
- ``GET  /api/sources``        → 200, owner-scoped list, newest-first (auth).
- ``GET  /api/sources/{id}``   → 200 owner; 404 cross-user/missing (auth).
"""

from __future__ import annotations

import logging
from collections.abc import Sequence
from datetime import datetime
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, File, Form, Query, UploadFile, status
from pydantic import BaseModel

from app.application.corpus import ReadSection, ReadSourceStructure
from app.application.errors import StorageUnavailable
from app.application.sources import CreateSource, GetSource, ListSources
from app.domain.entities import (
    CorpusStructure,
    SectionContent,
    Source,
    StructureSection,
    User,
)
from app.infrastructure.web.csrf import enforce_csrf, enforce_origin
from app.infrastructure.web.dependencies import (
    AppSettings,
    get_authenticated_user,
    get_create_source,
    get_get_source,
    get_list_sources,
    get_read_section,
    get_read_source_structure,
)
from app.infrastructure.web.rate_limit import rate_limit_upload

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/sources", tags=["sources"])


class SourceSummary(BaseModel):
    """Public, secret-free view of a source (safe to return/log).

    Deliberately omits ``object_key`` and ``checksum`` — those are internal
    storage/integrity details, never exposed to clients (spec P1-Upload AC1).
    """

    id: UUID
    title: str
    filename: str
    byte_size: int
    content_type: str
    status: str
    created_at: datetime

    @classmethod
    def from_entity(cls, source: Source) -> SourceSummary:
        return cls(
            id=source.id,
            title=source.title,
            filename=source.filename,
            byte_size=source.byte_size,
            content_type=source.content_type,
            status=source.status,
            created_at=source.created_at,
        )


@router.post(
    "",
    status_code=status.HTTP_201_CREATED,
    dependencies=[
        Depends(rate_limit_upload),
        Depends(enforce_origin),
        Depends(enforce_csrf),
    ],
)
def create_source(
    user: Annotated[User, Depends(get_authenticated_user)],
    service: Annotated[CreateSource, Depends(get_create_source)],
    settings: AppSettings,
    file: Annotated[UploadFile, File()],
    title: Annotated[str, Form()],
) -> SourceSummary:
    """Store an uploaded EPUB or PDF and persist an owned source row (201 + summary).

    The read is bounded to ``max(epub_max_bytes, pdf_max_bytes) + 1`` so an oversize
    upload is detected (→ 413) without buffering more than the largest cap in the
    request worker; validation then enforces the exact per-format cap.
    """
    read_bound = max(settings.epub_max_bytes, settings.pdf_max_bytes) + 1
    data = file.file.read(read_bound)
    try:
        source = service(
            user=user,
            title=title,
            filename=file.filename or "",
            content_type=file.content_type or "",
            data=data,
        )
    except StorageUnavailable:
        # SRC-10: record the failure with the owner id only — no secrets, and no
        # row was persisted (the service raised before ``sources.add``).
        logger.warning("source upload storage failure", extra={"user_id": str(user.id)})
        raise
    logger.info(
        "source created", extra={"user_id": str(user.id), "source_id": str(source.id)}
    )
    return SourceSummary.from_entity(source)


@router.get("")
def list_sources(
    user: Annotated[User, Depends(get_authenticated_user)],
    service: Annotated[ListSources, Depends(get_list_sources)],
) -> list[SourceSummary]:
    """Return the caller's sources, newest-first (200; 401 if unauthenticated)."""
    return [SourceSummary.from_entity(s) for s in service(user=user)]


@router.get("/{source_id}")
def get_source(
    source_id: UUID,
    user: Annotated[User, Depends(get_authenticated_user)],
    service: Annotated[GetSource, Depends(get_get_source)],
) -> SourceSummary:
    """Return one owned source (200); 404 if missing or owned by another user."""
    return SourceSummary.from_entity(service(user=user, source_id=source_id))


class StructureSectionView(BaseModel):
    """One node in the nested section tree (CORP-11).

    ``children`` holds the sections nested beneath this one per the TOC hierarchy;
    the tree is built in this web layer from the flat depth-ordered read model, so
    the domain and repository stay flat (design §Tech Decisions).
    """

    title: str
    depth: int
    section_path: list[str]
    anchor: str
    children: list[StructureSectionView]


class BookStructureView(BaseModel):
    """Public view of a source's parsed book structure (CORP-11).

    Book metadata plus the nested section tree. ``title``/``language`` are null and
    ``authors`` empty when the OPF omitted them (CORP-01).
    """

    title: str | None
    authors: list[str]
    language: str | None
    sections: list[StructureSectionView]

    @classmethod
    def from_structure(cls, structure: CorpusStructure) -> BookStructureView:
        return cls(
            title=structure.title,
            authors=list(structure.authors),
            language=structure.language,
            sections=_nest_sections(structure.sections),
        )


def _nest_sections(sections: Sequence[StructureSection]) -> list[StructureSectionView]:
    """Fold the flat, depth/position-ordered sections into a TOC tree.

    Each section nests under the most recent preceding section with a smaller
    depth (its parent); sections with no such ancestor are roots.
    """
    roots: list[StructureSectionView] = []
    stack: list[StructureSectionView] = []
    for section in sections:
        node = StructureSectionView(
            title=section.title,
            depth=section.depth,
            section_path=list(section.section_path),
            anchor=section.anchor,
            children=[],
        )
        while stack and stack[-1].depth >= section.depth:
            stack.pop()
        if stack:
            stack[-1].children.append(node)
        else:
            roots.append(node)
        stack.append(node)
    return roots


@router.get("/{source_id}/structure")
def get_source_structure(
    source_id: UUID,
    user: Annotated[User, Depends(get_authenticated_user)],
    service: Annotated[ReadSourceStructure, Depends(get_read_source_structure)],
) -> BookStructureView:
    """Return the owner's parsed book structure (200); 404 missing/non-owner/no-corpus."""
    return BookStructureView.from_structure(service(user=user, source_id=source_id))


class SectionContentView(BaseModel):
    """Public view of one section's readable content (FE-14).

    ``anchor`` echoes the requested location so a citation round-trips to exactly
    this section; ``section_path`` is the root-to-node breadcrumb and ``markdown``
    is the section's derived text for the reader.
    """

    anchor: str
    title: str
    section_path: list[str]
    markdown: str

    @classmethod
    def from_content(cls, section: SectionContent) -> SectionContentView:
        return cls(
            anchor=section.anchor,
            title=section.title,
            section_path=list(section.section_path),
            markdown=section.markdown,
        )


@router.get("/{source_id}/section")
def get_source_section(
    source_id: UUID,
    anchor: Annotated[str, Query(min_length=1)],
    user: Annotated[User, Depends(get_authenticated_user)],
    service: Annotated[ReadSection, Depends(get_read_section)],
) -> SectionContentView:
    """Return one owned section by anchor (200); 404 missing/non-owner/no-corpus/unknown-anchor."""
    return SectionContentView.from_content(
        service(user=user, source_id=source_id, anchor=anchor)
    )
